File: sdk/py/tutorials/ml/nlp/tu_nlp_bpe_transformer.py
#!/usr/bin/env python3
"""OA Tutorial — BPE Transformer · all-position LM (autograd).

Mirrors sdk/rs/tutorials/ml/nlp/tu_nlp_bpe_transformer.rs.
"""

import sys
sys.path.insert(0, str(__import__("pathlib").Path(__file__).parent))
import _nlp_common as nlp
import oa


def main() -> None:
	engine = oa.Engine()
	tokenizer = nlp.build_bpe_tokenizer()
	model = oa.ml.BpeTransformer(engine)
	optimizer = oa.ml.AdamW(model.all_parameters(), learning_rate=0.01)
	nlp.run_bpe("BPE Transformer · all-position LM", model, optimizer, tokenizer, engine)


if __name__ == "__main__":
	main()

File: sdk/rs/tutorials/ml/nlp/tu_nlp_bpe_transformer.rs
mod _bpe_common;

fn main() -> oa::Result<()> {
	_bpe_common::run(
		_bpe_common::Tutorial {
			title: "OA Tutorial — BPE Transformer · all-position LM (autograd)",
			description: "BPE + position Embedding → Transformer(32,64) → LN → Linear(32→320)",
			timer_name: "bpe_transformer_training_step",
			checkpoint_stem: "oars_bpe_transformer",
			learning_rate: 0.01,
			expected_loss: oa::sdk::ml::nlp::BPE_TRANSFORMER_FINAL_LOSS,
			expected_accuracy: oa::sdk::ml::nlp::BPE_TRANSFORMER_ACCURACY,
			expected_generation: oa::sdk::ml::nlp::BPE_TRANSFORMER_REFERENCE_GENERATION,
		},
		oa::sdk::ml::nlp::BpeTransformer::new,
	)
}
